### PR TITLE
feat(pegboard): add process monitoring, wip sqlite

### PR DESCRIPTION
**Pull request by MasterPtato**
Fixes RVTEE-608

### CODE REVIEW

**Comment by MasterPtato** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/1146?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#1146** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/1146?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#1145** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/1145?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1144** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/1144?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1143** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/1143?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1142** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/1142?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1141** <a href="https://app.graphite.dev/github/pr/rivet-gg/rivet/1141?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @MasterPtato and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: lib/pegboard/manager/src/utils.rs [hunk 1]
```diff
@@ -43,3 +83,167 @@ pub fn get_s3_path_stub(url: &Url, with_uuid: bool) -> Result<PathBuf> {
...
+			// `Create` event is received in wait_for_creation
+			let _ = tx.blocking_send(res);
+		},
+		Config::default().with_poll_interval(Duration::from_secs(2)),
+	)?;
+
+	Ok((watcher, rx))
+}
+
+pub async fn wait_for_creation<P: AsRef<Path>>(path: P) -> Result<()> {
```
Comment (NathanFlurry) [inline lib/pegboard/manager/src/utils.rs:222]: make sure the systemd service has enough watchers configured. write the calculation in the comments (i.e. `supports X containers * Y watchers per container`) and comment in this func to update the systemd watcher count too

File: lib/pegboard/manager/src/utils.rs [hunk 2]
```diff
@@ -43,3 +83,167 @@ pub fn get_s3_path_stub(url: &Url, with_uuid: bool) -> Result<PathBuf> {
...
+	let mut i = 0;
+
+	loop {
+		match cb().await {
+			std::result::Result::Ok(x) => return Ok(x),
+			std::result::Result::Err(err) => {
+				use sqlx::Error::*;
+
+				if i > MAX_QUERY_RETRIES {
+					bail!("max sql retries: {err}");
```
Comment (NathanFlurry) [inline lib/pegboard/manager/src/utils.rs:166]: where do these errors get written & monitored? we need to be able to catch every tiny issue with pb monitors

File: lib/pegboard/manager/src/ctx.rs [hunk 3]
```diff
@@ -1,23 +1,201 @@
...
+		// Attempt to use an existing connection
+		if let Some(conn) = self.pool.try_acquire() {
+			std::result::Result::Ok(conn)
+		} else {
+			// Create a new connection
+			self.pool.acquire().await.map_err(Into::into)
+		}
+	}
+
+	async fn send_packet(&self, packet: protocol::ToServer) -> Result<()> {
```
Comment (NathanFlurry) [inline lib/pegboard/manager/src/ctx.rs:54]: what happens on failure? how are all socket close edge cases handled/

File: lib/pegboard/manager/src/ctx.rs [hunk 4]
```diff
@@ -1,23 +1,201 @@
...
+		let buf = packet.serialize()?;
+		self.tx.lock().await.send(Message::Binary(buf)).await?;
+
+		Ok(())
+	}
+
+	async fn write_event(&self, event: &protocol::Event) -> Result<()> {
+		// Write event to db
+		let event_json = serde_json::to_vec(event)?;
+		utils::query(|| async {
```
Comment (NathanFlurry) [inline lib/pegboard/manager/src/ctx.rs:64]: how is this edge case failure handled?

File: lib/pegboard/manager/src/main.rs [hunk 5]
```diff
@@ -1,35 +1,63 @@
...
 mod container;
 mod ctx;
 mod utils;
 
-use client::Client;
+use ctx::Ctx;
+
+const PROTOCOL_VERSION: u16 = 1;
 
 #[tokio::main]
```
Comment (NathanFlurry) [inline lib/pegboard/manager/src/main.rs:17]: should we use a current threaded scheduler or limit the threads to 2? this is currently multithreaded, will spin up a lot of excess threads.

Comment (NathanFlurry) [reply lib/pegboard/manager/src/main.rs:17]: we do the same in the toolchain

Comment (NathanFlurry) [reply lib/pegboard/manager/src/main.rs:17]: idk how much parallelism we'll actually need

**Review by NathanFlurry — CHANGES_REQUESTED** [review]

File: lib/pegboard/manager/src/utils.rs (hunk 2 above)
Comment (MasterPtato) [reply lib/pegboard/manager/src/utils.rs:166]: Was planning on having vector pick up the logs from the manager and stream them to the backend. Didn't get to it for MVP I don't think

File: lib/pegboard/manager/src/ctx.rs (hunk 4 above)
Comment (MasterPtato) [reply lib/pegboard/manager/src/ctx.rs:64]: which failure? the serde_json::to_vec?

File: lib/pegboard/manager/src/ctx.rs (hunk 3 above)
Comment (MasterPtato) [reply lib/pegboard/manager/src/ctx.rs:54]: the manager errors then restarts and reconnects to the ws

**Pull request closed without merging**